otx/api/utils/tiler.py

```diff
@@ -68,87 +100,98 @@ def predict(self, image: np.ndarray):
-        return detections, features
-
-    def resize_masks(self, masks: List, dets: np.ndarray, shape: List[int]):
-        """Resize Masks.
+    def predict_sync(self, image: np.ndarray):
+        """Predict by cropping full image to tiles synchronously.
 
         Args:
-            masks (List): list of raw np.ndarray masks
-            dets (np.ndarray): detections including labels, scores, and boxes
-            shape (List[int]): original full-res image shape
+            image (np.ndarray): full size image
+
+        Returns:
+            detection: prediction results
+            features: saliency map and feature vector
         """
-        for i, (det, mask) in enumerate(zip(dets, masks)):
-            masks[i] = self.model.segm_postprocess(det[2:], mask, *shape[:-1])
+        features = (None, None)
+        tile_results = []
+        tile_coords = self.tile(image)
+        if isinstance(self.classifier, Model):
+            tile_coords = self.filter_tiles_by_objectness(image, tile_coords)
 
-    def predict_tile(
-        self,
-        image: np.ndarray,
-        coord: List[int],
-        masks: List[np.ndarray],
-        return_features=False,
-    ):
-        """Predict on single tile.
+        for i, coord in enumerate(tile_coords):
+            tile_img = self.crop_tile(image, coord)
+            tile_dict, tile_meta = self.model.preprocess(tile_img)
+            raw_predictions = self.model.infer_sync(tile_dict)
+            predictions = self.model.postprocess(raw_predictions, tile_meta)
+            tile_result = self.postprocess_tile(predictions, *coord[:2])
+            # cache full image feature vector and saliency map at 0 index
+            if i == 0 and ("feature_vector" in raw_predictions or "saliency_map" in raw_predictions):
+                features = (
+                    copy.deepcopy(raw_predictions["feature_vector"].reshape(-1)),
+                    copy.deepcopy(raw_predictions["saliency_map"][0]),
+                )
+
+            tile_results.append(tile_result)
+
+        results = self.merge_results(tile_results, image.shape)
+        return results, features
+
+    def predict_async(self, image: np.ndarray):
+        """Predict by cropping full image to tiles asynchronously.
 
         Args:
-            image (np.ndarray): full-res image
-            coord (List): tile coordinates
-            masks (List): list of raw np.ndarray masks
-            return_features (bool, optional): return saliency map and feature vector if set to true. Defaults to False.
+            image (np.ndarray): full size image
 
         Returns:
+            detection: prediction results
             features: saliency map and feature vector
-            output: single tile prediction
         """
+        tile_coords = self.tile(image)
+        if isinstance(self.classifier, Model):
+            tile_coords = self.filter_tiles_by_objectness(image, tile_coords)
```

Comment:
I think these lines are duplicated with `predict_sync`. Filtering tiles from tiling classifier uses `infer_sync`, which doesn't seems appropriate inside `predict_async`. I don't know well but if possible, it'd be better to move them into `predict` and then branch off into each `predict_sync` and `predict_async` with input of tile_coords. If not possible and hard, it's fine to let as is.

Reply:
Thank you for your feedback. I agree that it'd be ideal to have the code structured in the way you suggested, but as you mentioned, it may not be practical to implement at this time and I need some time to think about the overall design. I appreciate your suggestion and I will add a TODO in the code to revisit this in the future when feasible.

Reply:
Thanks! That's great advice. I've refactored accordingly.